clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java

```diff
@@ -273,9 +310,18 @@ void cleanup() {
             log.error("Unexpected error during shutdown.  Proceed with closing.", e);
         } finally {
             sendUnsentRequests(timer);
+
+            LinkedList<ApplicationEvent> allEvents = new LinkedList<>();
+            applicationEventQueue.drainTo(allEvents);
+            List<CompletableApplicationEvent<?>> completableEvents = allEvents
+                    .stream()
+                    .filter(e -> e instanceof CompletableApplicationEvent<?>)
+                    .map(e -> (CompletableApplicationEvent<?>) e)
+                    .collect(Collectors.toList());
```

Comment:
This logic is always needed whenever we `reapIncomplete`, and is currently repeated when we call it from the AsyncConsumer or here, so what about we move it into the `reapIncomplete`, make it receive a list of all events and internally filter the ones that are `CompletableEvent`

Reply:
😆 Yes, I went back and forth on this at least three times during development. I'll look at switching back to the approach you suggest.